File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    int target;
    int[] nums;
    List<List<Integer>> result;

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        this.target = target;
        this.nums = nums;
        this.result = new ArrayList<>();

        List<Integer> cur = new ArrayList<>();
        dfs(0, cur, 0);

        return this.result;
    }

    private void dfs(int i, List<Integer> cur, int total) {
        if (total == this.target) {
            result.add(new ArrayList<>(cur));
            return;
        }
        if (i >= this.nums.length || total > this.target) return;

        cur.add(this.nums[i]);
        dfs(i, cur, total + this.nums[i]);

        cur.remove(cur.size() - 1);
        dfs(i+1, cur, total);
    }
}

File: Data Structures & Algorithms/redundant-connection/submission-0.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        adj = {}
        for u, v in edges:
            if u not in adj:
                adj[u] = []
            if v not in adj:
                adj[v] = []
            
            visited = set()
            if self.dfs(adj, u, v, visited, -1) == True:
                return [u,v]

            adj[u].append(v)
            adj[v].append(u)

    def dfs(self, adj, cur: int, target: int, visited: set, parent: int) -> bool:
        if cur == target:
            return True

        visited.add(cur)

        if cur in adj:
            for i in adj[cur]:
                if i == parent or i in visited:
                    continue

                if self.dfs(adj, i, target, visited, cur):
                    return True

        return False
